# Titanic Survival Prediction: 02 - Modeling
*Date: 14.09.2025*
*Author: Jonas Lilletvedt*

--- 

## 1. Introduction

### 1.1. Objective

This notebook marks the second part of the Titanic Survival Prediction project: model training and evaluation. We will build directly on the pre-processing pipeline constructed in `01_data_cleaning_and_feature_engineering.ipynb`. The primary objective is to systematically train, evaluate, and tune a selection of models to find the optimal balance between predictive performance and interpretability. 

1. **Preparation and Pipeline Reconstruction:** We will begin by loading datasets and rebuilding the pre-processing pipeline from the previous notebook to ensure a consistent and reproducible environment.
2. **Baseline Model Evaluation:** A selection of diverse classification models (e.g., K-NN, Random Forest, Gradient Boosting) will be used as baseline and later help us identify strategies for optimization. 
3. **Hyperparameter Tuning:** The best-performing model from the baseline evaluation will undergo systematic hyperparameter tuning using GridSearchCV. The goal is to maximize its predictive performance based on our primary evaluation metric, the F1-Score. The reasoning for selecting this metric is detailed in the section below.
4. **Final Analysis and Iteration:** Finally, we will analyze the results and train the optimized model on the full dataset to generate final predictions for a Kaggle submission. We will conclude by exploring potential avenues for future improvements, such as alternative modeling techniques or further feature engineering.

Our end goal is twofold: to develop a model that is competitive on the Kaggle leaderboard, while remaining understandable enough to provide insights into the factors that influence survival. The process will be iterative, allowing us to refine our approach based on model performance. 

Beyond the Leaderboard: A Focus on Interpretability
Unlike many competition-driven projects that prioritize raw predictive power above all else, this analysis places a strong emphasis on interpretability. While achieving a high Kaggle score is a key objective, it is equally important to build a model whose decision-making process can be understood. By favoring models and features that provide clear insights—for instance, by showing why certain passengers were more likely to survive—we aim to produce not just a "black box" predictor, but a meaningful analysis of a historical event. This balanced approach ensures that our final result is both powerful and insightful.

### 1.2 Defining Success: Beyond Simple Accuracy

While accuracy is a straightforward metric, it can be misleading. A simple "naive" model that predicts survival for all females and death for all males achieves an accuracy of approximately 78.7% on the training data. Any model we build must therefore significantly outperform this baseline to be considered valuable.

To truly understand our model's performance, we must select the right evaluation metrics for the task. The most useful evaluation metric depends on the specific problem you are trying to solve. For example, in medical screening for a virus, achieving a high Recall is more important than high Precision. We would want to identify as many infected people as possible, even if it means accepting some false positives. In other situations, however, overall Accuracy might be the primary goal.
Applying this to the Titanic problem, our context is one of historical analysis and balanced prediction. There is no real-world cost that makes predicting a death incorrectly worse than predicting a survival incorrectly. Our goal is not just to be right, but to build a model that is equally effective at identifying both those who survived and those who did not.
For this reason, we need metrics that reward this balance and are not skewed by the class distribution. While we will still consider overall Accuracy, our primary metric for model evaluation is:
*   F1-Score: The harmonic mean of precision and recall. Provides an overall score between precision and recall. 

## 2. Data Loading and Setup

---

### 2.1. Library Imports and Seed

In [136]:
# Import necessary libraries

# Data manipulation and analysis
import numpy as np
import pandas as pd

# Set seed for reproducible results
seed = 43

### 2.2. Load Datasets

In [137]:
# Load datasets
df_train = pd.read_csv('../data/01_raw/train.csv')
df_test = pd.read_csv('../data/01_raw/test.csv')

### 2.3. Initial Inspection

A quick inspection to check the datasets are loaded properly and as expected.

**Datasets Shape:**

In [138]:
# Check shape of each dataset
print(f'Training data shape: {df_train.shape}')
print(f'Test data shape: {df_test.shape}')

Training data shape: (891, 12)
Test data shape: (418, 11)


**Data Preview:**

In [139]:
# Check five first rows in df_train
df_train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [140]:
# Check five first rows in df_test
df_test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

### 2.4. Prepare Data for Modeling

The datasets are loaded as expected. We will now separate our training data into two distinct objects:
*   **X:** A DataFrame containing all predictor variables.
*   **y:** A Series containing the target variable `Survived`, which we aim to predict.

In [141]:
# Seperate predictors from target
X_train = df_train.drop('Survived', axis=1)
y_train = df_train['Survived']

To avoid making changes on the original test data we will copy it to a separate object.

In [142]:
# Copy to avoid making changes to the original dataset
X_test = df_test.copy()

Display the shapes of our new variables to ensure the separation and copy worked as intended.

In [143]:
print(f'Shape of X_train: {X_train.shape}')
print(f'Shape of y_train: {y_train.shape}')
print(f'Shape of X_test: {X_test.shape}')

Shape of X_train: (891, 11)
Shape of y_train: (891,)
Shape of X_test: (418, 11)


### 2.5. Reconstruct the Pre-processing Pipeline

To make this notebook self-contained and reproducible, we will now redefine the custom transformers and the full `grand_pipeline` that were built and validated in the previous notebook. All the data transformation logic is encapsulated within a single code cell below. 

**A Note on Modularity vs. Narrative Flow**

In a production-level software project, this logic would typically be defined in a separate python-file to be imported. However, for this project, which is designed to be a linear, narrative-driven analysis, I have chosen to explicitly include the code here. This approach ensures that the notebook tells the complete, end-to-end story. Any future changes or improvements made in later stages of the project does not affect the logic or result of previous notebooks, preserving the integrity of each step of of our analysis. Furthermore, it allows any modifications or improvements to be documented and explained at the precise moment they are introduced, for a more natural progression.

All data transformation logic is therefore encapsulated within the single code cell below, which can be collapsed for easier reading of the modeling work that follows.

In [144]:
# Scikit-learn tools for preprocessing and modeling
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, FunctionTransformer
from sklearn.base import BaseEstimator, TransformerMixin

class TitleExtractor(BaseEstimator,  TransformerMixin):
    def __init__(self, rare_threshold=10):
        self.rare_threshold = rare_threshold

        # Titles with same meaning
        self.title_synonym_mapping_ = {
            'Mlle.': 'Miss.',
            'Ms.': 'Miss.',
            'Mme.': 'Mrs.'
        }

    def fit(self, X, y=None):
        titles = X['Name'].str.extract(pat=' ([A-Za-z]+\.)', expand=False)
        titles = titles.replace(self.title_synonym_mapping_)
        self.non_rare_titles_ = titles.value_counts()[lambda x: x >= self.rare_threshold].index
        return self
    
    def transform(self, X, y=None):
        # Copy to avoid modifying original
        X_copy = X.copy()
        # Extract title 
        X_copy['Title_feat'] = X_copy['Name'].str.extract(pat=' ([A-Za-z]+\.)', expand=False)
        
        # Swap titles with 'Rare' and synonym
        X_copy['Title_feat'] = X_copy['Title_feat'].replace(self.title_synonym_mapping_)


        X_copy['Title_feat'] = X_copy['Title_feat'].apply(lambda x: x if x in self.non_rare_titles_ else 'Rare')

        return X_copy

class FamilySurvialRateExtractor(BaseEstimator, TransformerMixin):
    def __init__(self, drop_surname=True, smooth_factor=1):
        self.drop_surname = drop_surname
        self.smooth_factor = smooth_factor

    def fit(self, X, y=None):
        if y is None:
            raise ValueError('''The FamilySurvialRateExtractor requires target variable 'y' for fitting''')

        X_temp = pd.concat([X, y], axis=1)
        X_temp['FamilyID_temp'] = self.__get_family_id(X)


        self.family_stats_ = X_temp.groupby('FamilyID_temp').agg(
            FamilySize_temp = ('Survived', 'size'), 
            FamilySurvivalCount_temp = ('Survived', 'sum')
        )
    
        self.global_survival_rate_ = y.mean()
        self.training_index_ = X.index
        self.y_train_ = y

        solo_travelers_stats = self.family_stats_[self.family_stats_['FamilySize_temp'] == 1]
        
        if not solo_travelers_stats.empty:
            self.alone_survival_rate_ = solo_travelers_stats['FamilySurvivalCount_temp'].mean()
        else:
            self.alone_survival_rate_ = self.global_survival_rate_

        return self

    def transform(self, X, y=None):
        X_copy = X.copy()

        X_copy['FamilyID_temp'] = self.__get_family_id(X)

        X_copy = X_copy.merge(self.family_stats_, on='FamilyID_temp', how='left')

        # Different calculation for passengers from training set -- avoid data leakage
        if X.index.equals(self.training_index_):
            X_copy['FamilySurvivalCount_temp'] -= self.y_train_
            X_copy['FamilySize_temp'] -= 1

        numerator = X_copy['FamilySurvivalCount_temp'] 
        denominator = X_copy['FamilySize_temp']
        
        # Apply smoothing
        smoothed_numerator = numerator + (self.smooth_factor * self.global_survival_rate_)
        smoothed_denominator = denominator + self.smooth_factor

        # Calculate FamilySurvivalRate
        X_copy['FamilySurvivalRate_feat'] = smoothed_numerator / smoothed_denominator

        # For passengers without a family from training
        X_copy['FamilySurvivalRate_feat'] = X_copy['FamilySurvivalRate_feat'].fillna(self.global_survival_rate_)

        # For passengers which travel alone we will overwrite global_survival_rate_
        is_alone_mask = (X_copy.groupby('FamilyID_temp')['FamilyID_temp'].transform('count') == 1) & (X_copy['FamilySize_temp'] == 0 | X_copy['FamilySize_temp'].isna())

        X_copy.loc[is_alone_mask, 'FamilySurvivalRate_feat'] = self.alone_survival_rate_

        # Columns to drop
        columns_to_drop = ['FamilySurvivalCount_temp', 'FamilySize_temp', 'FamilyID_temp']
        
        # Clean df
        X_copy = X_copy.drop(columns_to_drop, axis=1)

        return X_copy

    def __get_surname(self, X):
        # Extract surname
        data = X.copy()
        return data['Name'].str.extract(pat=r'^(.+)?,', expand=False)
    
    def __get_family_id(self, X):
        return self.__get_surname(X) + '_' + X['Pclass'].astype(str)

class AgeImputer(BaseEstimator, TransformerMixin):
    def __init__(self):
        pass

    def fit(self, X, y=None):
        X_temp = X.copy()
        self.median_by_title_ = X_temp.groupby('Title_feat')['Age'].median()
        return self
    
    def transform(self, X, y=None):
        X_copy = X.copy()
        X_copy['Age'] = X_copy['Age'].fillna(X_copy['Title_feat'].map(self.median_by_title_))
        return X_copy

class CabinLocationExtractor(BaseEstimator,  TransformerMixin):
    def __init__(self, drop_original=True):
        self.drop_original = True
    
    def fit(self, X, y=None):
        return self

    def transform(self, X, y=None):
        X_copy = X.copy()
        X_copy['Deck_feat'] = X_copy['Cabin'].str.extract(pat=r'^([A-Za-z])?', expand=False).fillna('U')
        X_copy['Zone_feat'] = pd.to_numeric(
            X_copy['Cabin'].str.extract(pat=r'([0-9]+)', expand=False),
            errors='coerce'
            )
        
        if self.drop_original:
            X_copy = X_copy.drop('Cabin', axis=1)

        return X_copy
    
feature_engineering_pipeline = Pipeline(steps=[
    ('family_survival_rate_extractor', FamilySurvialRateExtractor()),
    ('title_extractor', TitleExtractor()),
    ('age_imputer', AgeImputer()),
    ('cabin_location_extractor', CabinLocationExtractor())
])

FARE_TO_LOG_TRANS = ['Fare']
FAMILY_SURVIVAL_RATE_FEAT = ['FamilySurvivalRate_feat']
CAT_FEATURES = ['Sex',
                        'Embarked',
                        'Pclass',
                        'Title_feat',
                        'Deck_feat'
                        ]
AGE_TO_BIN = ['Age']
ZONE_TO_BIN = ['Zone_feat']


fare_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('log_transform', FunctionTransformer(np.log1p)),
    ('scaler', StandardScaler())
])

family_survival_rate_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

categorical_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore', sparse_output=False))
])

# Simple function for pd.cut
def apply_pd_cut(X, bins, labels):
    series = pd.Series(np.array(X).ravel()) 
    binned_series = pd.cut(series, bins=bins, labels=labels, right=True, include_lowest=True)
    return binned_series.to_numpy().reshape(-1, 1)

# Bins for Age
# Infant: 0-5, Child: 6-12, young-adult: 13-25, adult: 26-50, elder: 51->
AGE_BINS = [0, 5, 12, 25, 50, np.inf]
AGE_LABELS = ['Infant', 'Child', 'Young-Adult', 'Adult', 'Senior']

age_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('binner', FunctionTransformer(apply_pd_cut, kw_args={'bins': AGE_BINS, 'labels': AGE_LABELS})),
    ('onehot', OneHotEncoder(handle_unknown='ignore', sparse_output=False))
])

def apply_bin_zone(X, n_bins):
    series = pd.Series(np.array(X).ravel()) 
    labels  = [f'Q{i}' for i in range(1, n_bins + 1)]
    binned_series = pd.qcut(series, q=n_bins, labels=labels, duplicates='drop')

    # Convert nans to unknown
    binned_series = binned_series.cat.add_categories(['Unknown']).fillna('Unknown')
    return binned_series.to_numpy().reshape(-1, 1)

zone_pipeline = Pipeline(steps=[
    ('bin_with_missing_values', FunctionTransformer(apply_bin_zone, kw_args={'n_bins': 8})),
    ('onehot', OneHotEncoder(handle_unknown='ignore', sparse_output=False))
])

preprocessor = ColumnTransformer(
    transformers=[
        ('fare_log', fare_pipeline, FARE_TO_LOG_TRANS),
        ('family_survival_rate', family_survival_rate_pipeline, FAMILY_SURVIVAL_RATE_FEAT),
        ('cat', categorical_pipeline, CAT_FEATURES),
        ('age_binned', age_pipeline, AGE_TO_BIN),
        ('zone_binned', zone_pipeline, ZONE_TO_BIN)
    ],
    remainder='drop'
)

grand_pipeline = Pipeline(steps=[
    ('feature_engineering', feature_engineering_pipeline),
    ('preprocessing', preprocessor)
])  


## 3. Baseline Model Evaluation

---
With our pre-processing pipeline fully constructed, we can now proceed to the modeling phase. The first step is to establish a performance baseline by evaluating several different classification algorithms. This process will help us identify the most promising model architecture before investing time in hyperparameter tuning.

As established in the introduction, our primary metric for model selection will be the F1-Score, as it provides a balanced measure of a model's performance. For a more complete picture, we will also evaluate Precision, Recall, and overall Accuracy.

To obtain a reliable estimate of each model's generalization performance, we will use 5-fold cross-validation. This method helps mitigate the risk of overfitting to a single train-test split and provides a more robust foundation for our subsequent decisions.

For our baseline we will use four different classification models. 
1. **Logistic Regression**
2. **K-Nearest Neighbors**
3. **Random Forest**
4. **Gradient Boosting**
5. **Support Vector Machine**

In [ ]:
# Removing Warnings -- Make output hard to read
# We get a future warning from pipeline in the code block below. We have used the pipeline and functions as intended, I am therefore choosing to ignore them.
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

In [146]:
# Import SKlearn models
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score, cross_validate
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.svm import SVC

# Define models
models = {
    'Logistic Regression': LogisticRegression(random_state=seed, max_iter=1000),
    'K-Nearest Neighbors': KNeighborsClassifier(),
    'Random Forest': RandomForestClassifier(random_state=seed),
    'Gradient Boosting': GradientBoostingClassifier(random_state=seed), 
    'Support Vector Machine': SVC(random_state=seed, probability=True)
}

scoring_metrics = ['f1', 'accuracy']

# Dict for results
results = {}

print('Starting baseline-evaluation of models...\n')

# Iterate over all models
for model_name, model in models.items():

    # Train model using the pre-constructed pipeline
    full_pipeline = Pipeline(steps=[
        ('preprocessor', grand_pipeline),
        ('classifier', model)
    ])

    cv_results = cross_validate(
        full_pipeline, 
        X_train, 
        y_train, 
        cv=5, 
        scoring=scoring_metrics,
        n_jobs=1,  
        error_score='raise'
    )

    # Store and print the results for both metrics
    mean_f1 = np.mean(cv_results['test_f1'])
    std_f1 = np.std(cv_results['test_f1'])
    mean_accuracy = np.mean(cv_results['test_accuracy'])
    std_accuracy = np.std(cv_results['test_accuracy'])
    
    results[model_name] = {
        'mean_f1': mean_f1, 'std_f1': std_f1,
        'mean_accuracy': mean_accuracy, 'std_accuracy': std_accuracy
    }

    print(f'''{model_name}: Mean F1-Score = {results[model_name]['mean_f1']:.4f}''')
    print(f'''{model_name}: Std-F1-Score = {results[model_name]['std_f1']:.4f}''')
    print(f'''{model_name}: Mean Accuracy Score = {results[model_name]['mean_accuracy']:.4f}''')
    print(f'''{model_name}: Std-Accuracy Score = {results[model_name]['std_accuracy']:.4f}\n''')

best_model_name = max(results, key=lambda k: results[k]['mean_f1'])

print(f'''Best performing model: {best_model_name} with a mean F1-Score of {results[best_model_name]['mean_f1']:.4f}''')


Starting baseline-evaluation of models...

Logistic Regression: Mean F1-Score = 0.7617
Logistic Regression: Std-F1-Score = 0.0206
Logistic Regression: Mean Accuracy Score = 0.8193
Logistic Regression: Std-Accuracy Score = 0.0146

K-Nearest Neighbors: Mean F1-Score = 0.7431
K-Nearest Neighbors: Std-F1-Score = 0.0560
K-Nearest Neighbors: Mean Accuracy Score = 0.8137
K-Nearest Neighbors: Std-Accuracy Score = 0.0384

Random Forest: Mean F1-Score = 0.7528
Random Forest: Std-F1-Score = 0.0410
Random Forest: Mean Accuracy Score = 0.8205
Random Forest: Std-Accuracy Score = 0.0189

Gradient Boosting: Mean F1-Score = 0.7595
Gradient Boosting: Std-F1-Score = 0.0209
Gradient Boosting: Mean Accuracy Score = 0.8272
Gradient Boosting: Std-Accuracy Score = 0.0039

Support Vector Machine: Mean F1-Score = 0.7647
Support Vector Machine: Std-F1-Score = 0.0392
Support Vector Machine: Mean Accuracy Score = 0.8361
Support Vector Machine: Std-Accuracy Score = 0.0176

Best performing model: Support Vector Mach

The baseline evaluation identifies the **Support Vector Machine (SVM)** as the best performing model in both key metrics:

*   **Mean F1-Score:** 0.7647
*   **Mean Accuracy:** 0.8361


Gradient Boosting is a very close second (0.7595 F1-Score), both looks like promising candidates for further improvements.

For now we will move forward with SVM and Gradient Boosting as models of choice for hyperparameter tuning.

## 4. Hyperparameter Tuning

---

The next step is to tune the Support Vector Machine in hopes of further improving its performance. We will use GridSearchCV to systematically test different combinations of the model's key parameters to find the optimal combination for our data.

### 4.1. Hyperparameter Tuning Support Vector Machine

In [ ]:
from sklearn.model_selection import GridSearchCV

# Best performing model from base evaluation
svc = SVC(random_state=seed, probability=True)

# Create tuning pipeline
tuning_pipeline = Pipeline(steps=[
    ('preprocessor', grand_pipeline),
    ('classifier', svc)
])

# Define parameter grid
param_grid = {
    'classifier__C': [0.1, 1, 10, 50, 100],
    'classifier__gamma': ['scale', 'auto', 0.01, 0.1, 1], 
    'classifier__kernel': ['rbf']
}

# Create and run GridSearchCV object
grid_search = GridSearchCV(
    tuning_pipeline, 
    param_grid,
    cv=5, 
    scoring={'f1': 'f1', 'accuracy': 'accuracy'},
    refit='f1',
    n_jobs=1
)

# Terminal comments
print('Starting Hyperparameter Tuning for Support Vector Machine...')

# Run tuning
grid_search.fit(X_train, y_train)

best_model_index = grid_search.best_index_
best_accuracy = grid_search.cv_results_['mean_test_accuracy'][best_model_index]

# Display results
print('F1-Score Performance:')
print(f'''\n  - Baseline F1 score: {results['Support Vector Machine']['mean_f1']:.4f}''')
print(f'  - Tuned F1 score:    {grid_search.best_score_:.4f}')

print('\nAccuracy Performance:')
print(f'''  - Baseline Accuracy: {results['Support Vector Machine']['mean_accuracy']:.4f}''')
print(f'  - Accuracy of Tuned Model: {best_accuracy:.4f}')

final_model = grid_search.best_estimator_

Starting Hyperparameter Tuning for Support Vector Machine...
F1-Score Performance:

  - Baseline F1 score: 0.7647
  - Tuned F1 score:    0.7819

Accuracy Performance:
  - Baseline Accuracy: 0.8361
  - Accuracy of Tuned Model: 0.8384


## 4.2. Hyperparameter Tuning Gradient Boosting

In [ ]:
gb = GradientBoostingClassifier(random_state=seed)

# Create tuning pipeline
tuning_pipeline = Pipeline(steps=[
    ('preprocessor', grand_pipeline),
    ('classifier', gb)
])

# Define parameter grid
param_grid = {
    'classifier__n_estimators': [100, 200, 300],
    'classifier__learning_rate': [0.01, 0.05, 0.1],
    'classifier__max_depth': [3, 4, 5],
    'classifier__subsample': [0.7, 0.9, 1.0]
}

# Create and run GridSearchCV object
grid_search = GridSearchCV(
    tuning_pipeline, 
    param_grid,
    cv=5, 
    scoring={'f1': 'f1', 'accuracy': 'accuracy'},
    refit='f1',
    n_jobs=-1 # Takes so long with 1 -- Fuck me warnings
)

# Terminal comments
print('Starting Hyperparameter Tuning for Gradient Boosting...')

# Run tuning
grid_search.fit(X_train, y_train)

best_model_index = grid_search.best_index_
best_accuracy = grid_search.cv_results_['mean_test_accuracy'][best_model_index]

# Display results
print('F1-Score Performance:')
print(f'''\n  - Baseline F1 score: {results['Gradient Boosting']['mean_f1']:.4f}''')
print(f'  - Tuned F1 score:    {grid_search.best_score_:.4f}')

print('\nAccuracy Performance:')
print(f'''  - Baseline Accuracy: {results['Gradient Boosting']['mean_accuracy']:.4f}''')
print(f'  - Accuracy of Tuned Model: {best_accuracy:.4f}')

final_model = grid_search.best_estimator_

Starting Hyperparameter Tuning for Gradient Boosting...


/home/jonaslilletvedt/miniconda3/envs/titanic-survival-prediction/lib/python3.11/site-packages/sklearn/pipeline.py:62: FutureWarning: This Pipeline instance is not fitted yet. Call 'fit' with appropriate arguments before using other methods such as transform, predict, etc. This will raise an error in 1.8 instead of the current warning.
  warnings.warn(
/home/jonaslilletvedt/miniconda3/envs/titanic-survival-prediction/lib/python3.11/site-packages/sklearn/pipeline.py:62: FutureWarning: This Pipeline instance is not fitted yet. Call 'fit' with appropriate arguments before using other methods such as transform, predict, etc. This will raise an error in 1.8 instead of the current warning.
  warnings.warn(
/home/jonaslilletvedt/miniconda3/envs/titanic-survival-prediction/lib/python3.11/site-packages/sklearn/pipeline.py:62: FutureWarning: This Pipeline instance is not fitted yet. Call 'fit' with appropriate arguments before using other methods such as transform, predict, etc. This will raise 

F1-Score Performance:

  - Baseline F1 score: 0.7595
  - Tuned F1 score:    0.7775

Accuracy Performance:
  - Baseline Accuracy: 0.8272
  - Accuracy of Tuned Model: 0.8417


After hyperparameter tuning, both models showed marked improvement. The final cross-validation scores are as follows:

| Model                  | Metric     | Baseline Score | Tuned Score | Change     |
| :--------------------- | :--------- | :------------- | :---------- | :--------- |
| **Support Vector Machine** | **F1-Score** | **0.7647**     | **0.7819**  | **+0.0172**|
|                        | Accuracy   | 0.8361         | 0.8384      | +0.0023    |
| **Gradient Boosting**      | F1-Score   | 0.7595         | 0.7775      | +0.0180    |
|                        | **Accuracy**   | **0.8272**     | **0.8417**  | **+0.0145**|

### Key Findings

- The **Support Vector Machine (SVM)** achieved the highest F1-score, which was our primary metric for optimization.
- **Gradient Boosting** saw the largest overall improvement and now holds the highest accuracy.

While the tuned models granted some better results, I have two possible ideas to further improvement.

1. **XGBoost**: A more efficient and powerful implementation of Gradient Boosting, widely known as a go-to algorithm for machine learning competitions. It offers key advantages over the standard implementation, including advanced regularization to improve accuracy by preventing overfitting, and superior support for parallel processing to significantly reduce training times.

2. **Voting Classifier**: A Voting Classifier is an ensemble method that combines the predictions from multiple models to generate a more robust final prediction.In our case, we can take our two best-tuned models (Support Vector Machine and Gradient Boosting) and average their prediction probabilities. Different models learn in different ways and therefore make different types of errors. By combining them, we can often cancel out these individual mistakes, leading to a more accurate and reliable outcome. The key to a successful ensemble is model diversity, for this reason, we will combine our SVM and Gradient Boosting models, as they have fundamentally different structures. Combining two similar models, like Gradient Boosting and XGBoost, would be less effective as they are likely to make the same kinds of errors.